bin/aidiff_embeddings/export_google_drive_content.rb

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env ruby
+
+require_relative '../../dashboard/config/environment'
+require 'async'
+require 'async/barrier'
+require 'fileutils'
+require 'net/http'
+
+# Force very small downloads to be stored as Tempfiles instead of using StringIO objects
+OpenURI::Buffer.send :remove_const, :StringMax if OpenURI::Buffer.const_defined?(:StringMax)
+OpenURI::Buffer.const_set :StringMax, 0
+
+# These methods can/should be moved out of a monkey patch if/when this process
+# is moved out of an ad-hoc script
+# Regex based on /apps/src/templates/lessonOverview/googleDocsUtils.js.
+class Resource
+  GDOCS_URL_REGEX = /^https?:\/\/docs\.google\.com\/(document|presentation|spreadsheets)\/d\/([\w-]+)/
+
+  def google_docs?
+    !!(url =~ GDOCS_URL_REGEX)
+  end
+
+  def google_pdf_download_url
+    return nil unless google_docs?
+
+    url =~ GDOCS_URL_REGEX
+    "https://docs.google.com/#{$1}/d/#{$2}/export?format=pdf"
+  end
+end
+
+BASE_DIR = Dir.home + "/embeddings/"
+MAX_ATTEMPTS = 3
+
+# Mapping of download URLs to existing file paths. Many resources are referred
+# to from multiple units or lessons. This avoids downloading them multiple times.
+# Instead we make a separate copy locally with a distinct metadata file.
+$downloaded_files = {}
+
+def download_google_file(url:, path:, metadata:, barrier:)
+  barrier.async do
+    print '>'
+
+    full_path = BASE_DIR + path
+    existing_file = $downloaded_files[url]
+    if existing_file.present?
+      FileUtils.cp(existing_file, full_path)
+    else
+      begin
+        attempts ||= 1
+        handle = URI.parse(url).open
+      rescue => exception
+        # Some documents do not have public download permissions; skip these.
+        if exception&.io&.status.present?
+          status_code = exception.io.status[0]
+          next if status_code == '401'
+        end
+
+        puts "Network error:"
+        puts exception.inspect
+        if (attempts += 1) <= MAX_ATTEMPTS
+          puts "Retrying..."
+          retry
+        else
+          puts "Too many retries, skipping."
+          next
+        end
+      end
+
+      # TODO: Write these files to the S3 bucket instead of / in addition to local filesystem
+      FileUtils.cp(handle.path, full_path)
+      $downloaded_files[url] = full_path
+    end
+
+    full_metadata = {
+      metadataAttributes: metadata
+    }
+    File.write("#{full_path}.metadata.json", full_metadata.to_json)
+
+    print '<'
+  end
+end
+
+Async do
+  # Export resources from courses
+  UnitGroup.all.each do |course|
```

Comment:
the Jira issue says "all curriculum", and what you have here matches that, so that's good. However, I am curious if we really want this to run on `UnitGroup`s and `Unit`s which have `published_state` of `in_development`, or whose `participant_audience` is not `student`. Given the amount of data this copies, it could be nice not to have more than we need.

Reply:
Agree, I think we only need to have user-visible data in the knowledge base, but I think we might want differentiation for teacher courses as well? Idk, that seems like a @tess323 question

Reply:
A fine point. Before updating this script to write to S3 we'll discuss the set of resources we want to use, and I'll put those changes in place. I'll commit the current version as is.